### PR TITLE
Add tests for MP Config - Config registering and releasing

### PR DESCRIPTION
**Pull request by marekkopecky**
Add tests for MP Config - Config registering and releasing

https://issues.redhat.com/browse/WFLY-13023

Please make sure your PR meets the following requirements:
- [x] Pull Request contains a description of the changes
- [x] Pull Request does not include fixes for multiple issues/topics
- [x] Code is formatted, imports ordered, code compiles and tests are passing
- [ ] Link to the passing job is provided
- [x] Code is self-descriptive and/or documented
- [x] Description of the tests scenarios is included (see #46)

### CODE REVIEW

**Comment by honza-kasik** [comment]
@marekkopecky Can you please format the code and re-push? https://github.com/jboss-eap-qe/eap-microprofile-test-suite#code-format-and-imports-order

File: microprofile-config/src/test/java/org/jboss/eap/qe/microprofile/config/ResolverTest.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    /**
+     * @tpTestDetails Tests for ConfigProviderResolver#registerConfig` and ConfigProviderResolver#releaseConfig methods
+     * @tpPassCrit Verifies that the Config objects returns correct values with respect to registering and releasing default
+     *             or custom Config object
+     * @tpSince EAP 7.4.0.CD19
+     */
+    @Test
+    public void testAppEndpoint(@ArquillianResource URL baseURL) {
```
Comment (honza-kasik) [inline microprofile-config/src/test/java/org/jboss/eap/qe/microprofile/config/ResolverTest.java]: IMHO almost all tests could have similar name in this testsuite. Can you please rename it to be more specific?

File: microprofile-config/src/test/java/org/jboss/eap/qe/microprofile/config/ResolverTest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                        JaxRsTestApplication.class,
+                        ResolverEndPoint.class,
+                        ResolverConfigSource.class)
+                .addAsManifestResource(new StringAsset(String.format("%s=%s", PROPERTY_NAME, ORIGINAL_PROPERTY_VALUE)),
+                        "microprofile-config.properties")
+                .addAsWebInfResource(EmptyAsset.INSTANCE, "beans.xml");
+    }
+
+    /**
+     * @tpTestDetails Tests for ConfigProviderResolver#registerConfig` and ConfigProviderResolver#releaseConfig methods
```
[the remark this answers is not in the document]
Comment (marekkopecky) [reply microprofile-config/src/test/java/org/jboss/eap/qe/microprofile/config/ResolverTest.java]: +1, thank you for pointing this out, I will update it

File: microprofile-config/src/main/java/org/jboss/eap/qe/microprofile/config/testapp/jaxrs/ResolverEndPoint.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        // handle original config
+        Config config = resolver.getConfig();
+        response.append(config.getValue(PROPERTY_NAME, String.class));
+        resolver.releaseConfig(config);
+
+        // prepare, register and release new config
+        ConfigBuilder builder = resolver.getBuilder();
+        Config specificConfig = builder.addDefaultSources().withSources(new ResolverConfigSource()).build();
+        resolver.registerConfig(specificConfig, ResolverEndPoint.class.getClassLoader());
+        config = resolver.getConfig();
```
[the remark this answers is not in the document]
Comment (marekkopecky) [reply microprofile-config/src/main/java/org/jboss/eap/qe/microprofile/config/testapp/jaxrs/ResolverEndPoint.java]: +1, I will update this

File: microprofile-config/src/test/java/org/jboss/eap/qe/microprofile/config/ResolverTest.java (hunk 1 above)
Comment (marekkopecky) [reply microprofile-config/src/test/java/org/jboss/eap/qe/microprofile/config/ResolverTest.java]: +1

**Comment by marekkopecky** [comment]
@honza-kasik Thank you for your review, I updated PR accordingly.

**Comment by jmesnil** [comment]
I'm -1 on this test. with smallrye-config, we are challenging this feature in https://github.com/eclipse/microprofile-config/issues/136 as the `Config` lifecycle and ownership of resources is not properly specified.

Besides, the test is using SPI and API calls that should not be mixed in that fashion. (the Config API/SPI is poorly designed and there should have been a clearer split between those but the intent of MP Config is to have the SPI uses for integration/definition of Config and the API used by the application code to get config properties)

**Comment by marekkopecky** [comment]
@jmesnil Main point of https://github.com/eclipse/microprofile-config/issues/136 is to update Config live cycle with respect to `AutoCloseable` behaviour of `ConfigSource` and `Converter` classes.

Tests in this PR should cover new spec parts from MP Config 1.4 from this commit: https://github.com/eclipse/microprofile-config/commit/f228e42897adb721792bcf960cda22be95029f65#diff-b92b26a22047eb65f66c5f57bf42a049R53-R69
Specifically, these tests should cover registerConfig and releaseConfig methods. These methods can be used now (without additional clarification) according to the spec text. I'm OK with additional clarification of Config live cycle, but I believe that current behaviour should be tested now. I'm going to remove these tests from this MP TS once proper tests will be part of TCKs (this tests or different tests if https://github.com/eclipse/microprofile-config/issues/136  will be resolved.

File: microprofile-config/src/main/java/org/jboss/eap/qe/microprofile/config/testapp/jaxrs/ResolverEndPoint.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                return ILLEGAL_STATE_EXCEPTION_ERROR_MESSAGE;
+            }
+        }
+
+        // get value from first config
+        Config customConfig1 = resolver.getConfig();
+        response.append(customConfig1.getValue(PROPERTY_NAME, String.class));
+        resolver.releaseConfig(customConfig1);
+
+        // get value from second config
```
Comment (marekkopecky) [inline microprofile-config/src/main/java/org/jboss/eap/qe/microprofile/config/testapp/jaxrs/ResolverEndPoint.java]: @jmesnil Based on our f2f mtg, I'm adding second test that checks behaviour of two Config objects with the same ConfigSource object. This test checks that second Config object can't be registered by registerConfig. This test also checks property values from both Config objects.

Comment (marekkopecky) [reply microprofile-config/src/main/java/org/jboss/eap/qe/microprofile/config/testapp/jaxrs/ResolverEndPoint.java]: @jmesnil Are you ok with this PR now?

**Comment by honza-kasik** [comment]
Thank you @marekkopecky ! LGTM! Merging...

**Pull request merged**